### PR TITLE
Improved page numbering

### LINKED ISSUE #2873: Pageref in Typst like it exists in LaTeX
### Description

A possibility to get the page number when referencing an image or table would be good. Below is my own implementation that I use in Typst right now. It takes the label text and tries to find the label in the document. If the label is found it determines the page number of the label and formats it as a link to the label. In case the label is not defined it throws an error so it is not possible to compile the document with unresolved references when using pageref. 

```typst
#let pageref(labelText) = {
  locate(loc => {
    let label = label(labelText)
    let elems = query(selector(label), loc)

    if elems.len() == 1 {
      let pageNumber = elems.at(0).location().page()
      link(label, str(pageNumber))
    } else {
      let errorMessage = "Label '" + labelText + "' is not defined"
      panic(errorMessage)
    }
  })
}
```

### Use Case

Sometimes it can be handy to reference the page number in a document for example "As you can see in fig. x on page y".

### PR DESCRIPTION
**Pull request by Dherse**
Working towards closing #2873 by making the implementation of `counter(...).display()` better with the following changes:

- Numbering can now be specified using the `#[numbering]` attribute of the `elem` macro: this automatically sets the function pointer in `NativeElementData` to point to that numbering in the current stylechain
- `counter.display` can now take an optional location, that way instead of needing to do `numbering(..., counter(page).at(loc))` the user can now just do `counter.display(location: loc)` which imo is neater and easy to implement syntactic/functional sugar
- `counter.display` now obtains the correct numbering for the style chain for all elements that have the `#[numbering]` attribute.
- Reduced complexity in `outline.rs` since we can rely on the `.display` implementation to handle it for us instead of building the correct page numbering manually.

This is particularly useful for packages such as glossarium that rely on display the location of a page.

### CODE REVIEW

File: crates/typst-macros/src/elem.rs [hunk 1]
```diff
@@ -1301,3 +1333,30 @@ fn create_field_parser(field: &Field) -> (TokenStream, TokenStream) {
...
+    let idents_of_path = path.segments.iter().fold(String::new(), |mut acc, v| {
+        acc.push_str(&v.ident.to_string());
+        acc.push(':');
+        acc
+    });
+    ["Option:", "std:option:Option:", "core:option:Option:"]
+        .into_iter()
+        .find(|s| idents_of_path == *s)
+        .and_then(|_| path.segments.last())
+}
```
Comment (laurmaedje) [inline crates/typst-macros/src/elem.rs:1362]: This check seems overcomplicated, just a check for `Option` would suffice I think. But even if not, I don't think allocation is necessary.

File: crates/typst/src/introspection/counter.rs [hunk 2]
```diff
@@ -361,6 +360,12 @@ impl Counter {
         /// style exists.
         #[default]
         numbering: Option<Numbering>,
+        /// The location at which to display the counter. If omitted, displays
+        /// the counter at the location where the returned content is inserted
+        /// into the document. A suitable location can be retrieved from
+        /// [`locate`]($locate) or [`query`]($query).
+        #[default]
+        location: Option<Location>,
```
Comment (laurmaedje) [inline crates/typst/src/introspection/counter.rs:368]: another optional positional argument might make the API too confusing. I would also prefer to discuss this argument separately from the rest of the PR.

File: crates/typst/src/introspection/counter.rs [hunk 3]
```diff
@@ -643,15 +653,28 @@ impl Show for DisplayElem {
...
-                    if func == HeadingElem::elem() {
-                        HeadingElem::numbering_in(styles).clone()
-                    } else if func == FigureElem::elem() {
-                        FigureElem::numbering_in(styles)
-                    } else if func == EquationElem::elem() {
-                        EquationElem::numbering_in(styles)
-                    } else {
-                        None
+                    // Special case for page numbering if user sets the counter to:
+                    // `counter(page.before(loc))` or similar cases.
```
Comment (laurmaedje) [inline crates/typst/src/introspection/counter.rs:657]: The `.before(loc)` here is weird.

File: crates/typst/src/introspection/counter.rs [hunk 4]
```diff
@@ -643,15 +653,28 @@ impl Show for DisplayElem {
...
+                .or_else(|| {
+                    let CounterKey::Page = counter.0 else {
+                        return None;
+                    };
+
+                    engine
+                        .introspector
+                        .page_numbering(location)
+                        .cloned()
+                        .or_else(|| PageElem::numbering_in(styles).clone())
```
Comment (laurmaedje) [inline crates/typst/src/introspection/counter.rs:677]: Why is this duplicated?

**Comment by Dherse** [comment]
Closing this for now, I don't know if I'll get back to it. I will fix it and reopen it, unless somebody else wants to do it of course, when and if I find the motivation to. Sorry

**Pull request closed without merging**